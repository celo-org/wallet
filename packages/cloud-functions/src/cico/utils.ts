import * as admin from 'firebase-admin'
import { v4 as uuidv4 } from 'uuid'
import { bigQueryDataset, bigQueryProjectId, getBigQueryInstance } from '../bigQuery'

const bigQuery = getBigQueryInstance()

export interface UserDeviceInfo {
  id: string
  appVersion: string
  userAgent: string
}

interface UserInitData {
  ipAddress: string
  timestamp: string
  userAgent: string
}

export const getUserInitData = async (
  currentIpAddress: string,
  deviceId: string,
  userAgent: string
): Promise<UserInitData> => {
  const [data] = await bigQuery.query(`
    SELECT context_ip, device_info_user_agent, timestamp
    FROM ${bigQueryProjectId}.${bigQueryDataset}.app_launched
    WHERE user_address = (
        SELECT user_address
        FROM ${bigQueryProjectId}.${bigQueryDataset}.app_launched
        WHERE device_info_unique_id= "${deviceId}"
        AND user_address IS NOT NULL
        ORDER BY timestamp DESC
        LIMIT 1
    )
    ORDER BY timestamp ASC
    LIMIT 1
  `)

  if (!data.length) {
    return {
      ipAddress: currentIpAddress,
      timestamp: new Date().toISOString(),
      userAgent,
    }
  }

  const { context_ip, device_info_user_agent, timestamp } = data[0]

  const userInitData = {
    ipAddress: context_ip,
    timestamp: timestamp.value,
    userAgent: device_info_user_agent,
  }

  return userInitData
}

export const getOrCreateUuid = async (userAddress: string) => {
  let simplexId = await admin
    .database()
    .ref(`registrations/${userAddress}/simplexId`)
    .once('value')
    .then((snapshot) => snapshot.val())

  if (simplexId) {
    return simplexId
  }

  simplexId = uuidv4()
  await admin.database().ref(`registrations/${userAddress}`).update({ simplexId })

  return simplexId
}

export function getFirebaseAdminCreds(localAdmin: any) {
  if (!process.env.GCLOUD_PROJECT) {
    try {
      const serviceAccount = require('../../config/serviceAccountKey.json')
      return localAdmin.credential.cert(serviceAccount)
    } catch (error) {
      console.error(
        'Error: Could not initialize admin credentials. Is serviceAccountKey.json missing?',
        error
      )
    }
  } else {
    try {
      return localAdmin.credential.applicationDefault()
    } catch (error) {
      console.error('Error: Could not retrieve default app creds', error)
    }
  }
}

export const fetchWithTimeout = async (
  url: string,
  body: any | null,
  duration: number
): Promise<Response | undefined> => {
  try {
    // @ts-ignore
    const timeout = new Promise<undefined>((resolve, reject) => {
      const id = setTimeout(() => {
        clearTimeout(id)
        reject(Error(`Request timed out after ${duration}ms`))
      }, duration)
    })

    return Promise.race([body ? fetch(url, body) : fetch(url), timeout])
  } catch (error) {
    throw error
  }
}
