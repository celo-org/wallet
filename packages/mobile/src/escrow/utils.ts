import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToe164NumberSelector } from 'src/identity/reducer'
import { recipientCacheSelector } from 'src/recipients/reducer'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipientName = (payment: EscrowedPayment) => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToe164Number = useSelector(identifierToe164NumberSelector)
  const recipientCache = useSelector(recipientCacheSelector)
  const phoneNumber = identifierToe164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber]

  return recipient?.displayName || recipientPhone
}
