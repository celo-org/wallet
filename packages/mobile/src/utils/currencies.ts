export enum Currency {
  Celo = 'Celo Gold',
  Dollar = 'Celo Dollar',
  Euro = 'Celo Euro',
}

export interface CurrencyInfo {
  symbol: string
  code: string
  displayDecimals: number
}

type CurrencyObject = { [key in Currency]: CurrencyInfo }

export const CURRENCIES: CurrencyObject = {
  [Currency.Celo]: {
    symbol: '',
    code: 'cGLD',
    displayDecimals: 3,
  },
  [Currency.Dollar]: {
    symbol: '$',
    code: 'cUSD',
    displayDecimals: 2,
  },
  [Currency.Euro]: {
    symbol: 'â‚¬',
    code: 'cEUR',
    displayDecimals: 2,
  },
}

export const currencyByCode = (code: string): Currency => {
  for (const [currency, currencyInfo] of Object.entries(CURRENCIES)) {
    if (currencyInfo.code === code) {
      return currency as Currency
    }
  }
  return Currency.Dollar
}

export const resolveCurrency = (label: string): Currency => {
  if (label && label.toLowerCase().includes('dollar')) {
    return Currency.Dollar
  } else if (label && label.toLowerCase().includes('euro')) {
    return Currency.Euro
  } else if (label && label.toLowerCase().includes('gold')) {
    return Currency.Celo
  } else {
    console.info('Unable to resolve currency from label: ' + label)
    return Currency.Dollar
  }
}

export enum ShortCurrency {
  Dollar = 'dollar',
  Celo = 'gold',
  Euro = 'euro',
}

export const currencyToShortMap = {
  [Currency.Dollar]: ShortCurrency.Dollar,
  [Currency.Celo]: ShortCurrency.Celo,
  [Currency.Euro]: ShortCurrency.Euro,
}

export const STABLE_CURRENCIES = [Currency.Dollar, Currency.Euro]
