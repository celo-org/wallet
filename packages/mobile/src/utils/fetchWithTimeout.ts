export const fetchWithTimeout = async (
  url: string,
  body: any | null,
  duration: number
): Promise<Response> => {
  try {
    const timeout = new Promise<undefined>((resolve, reject) => {
      const id = setTimeout(() => {
        clearTimeout(id)
        reject(Error(`Request timed out after ${duration}ms`))
      }, duration)
    })

    const response = await Promise.race([body ? fetch(url, body) : fetch(url), timeout])
    // Response should always be defined because `reject` throws an error
    // but just satifying the linter with this check
    if (!response) {
      throw Error(`Request timed out after ${duration}ms`)
    }
    return response
  } catch (error) {
    throw error
  }
}
