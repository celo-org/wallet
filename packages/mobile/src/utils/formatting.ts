import colors from '@celo/react-components/styles/colors'
import BigNumber from 'bignumber.js'
import { CURRENCIES, CURRENCY_ENUM, WEI_PER_TOKEN } from 'src/geth/consts'
import { LocalCurrencyCode, LocalCurrencySymbol } from 'src/localCurrency/consts'

// Returns a localized string that represents the number with the right decimal points.
export const getMoneyDisplayValue = (
  value: BigNumber.Value,
  currency: CURRENCY_ENUM = CURRENCY_ENUM.DOLLAR,
  includeSymbol: boolean = false,
  roundingTolerance: number = 1
): string => {
  // TODO: use new CURRENCIES enum when cEUR work is merged
  const decimals = currency === CURRENCY_ENUM.GOLD ? 4 : CURRENCIES[currency].displayDecimals
  const symbol = CURRENCIES[currency].symbol
  const formattedValue = roundDown(value, decimals, roundingTolerance).toFormat(decimals)
  return includeSymbol ? symbol + formattedValue : formattedValue
}

export const getLocalCurrencyDisplayValue = (
  value: BigNumber.Value,
  currency: LocalCurrencyCode,
  includeSymbol: boolean = false,
  roundingTolerance: number = 1
): string => {
  const symbol = LocalCurrencySymbol[currency]
  const formattedValue = roundDown(value, 2, roundingTolerance).toFormat(2)
  return includeSymbol ? symbol + formattedValue : formattedValue
}

// like getMoneyDisplayValue but only returns cents if they are significant
export const getCentAwareMoneyDisplay = (value: BigNumber.Value): string => {
  const bigValue = new BigNumber(value)
  return bigValue.isInteger() ? bigValue.toFixed(0) : roundDown(value).toFormat(2)
}

export const getExchangeRateDisplayValue = (value: BigNumber.Value): string => {
  return new BigNumber(value).decimalPlaces(4).toFormat()
}

/**
 * Rounds up to upper-bound fees for display
 * Fails quietly (displays minimum value) on nonnegative inputs
 * @param value fee amount, assumed nonnegative
 * @param topLine true if display should be to two decimal places
 */
export const getFeeDisplayValue = (
  value: BigNumber.Value | null | undefined,
  topLine: boolean = true,
  isCELO: boolean
): string => {
  if (!value || new BigNumber(value).isZero()) {
    return '-'
  }
  const decimals = isCELO ? 4 : topLine ? 2 : 3
  const minDisplay = Math.pow(10, 0 - decimals)
  return value <= minDisplay
    ? new BigNumber(minDisplay).toFormat(decimals)
    : roundUp(value, decimals).toFormat(decimals)
}

/**
 * More precise getFeeDisplayValue with built in rounding
 * Used for small Network Fees
 * @param value fee amount
 * @param precise true if additional precision to 6 digits for <0.001 needed
 */
export const getNetworkFeeDisplayValue = (
  value: BigNumber.Value,
  precise: boolean = false
): string => {
  const roundedNumber = new BigNumber(value)
  if (precise && roundedNumber.isLessThan(0.000001)) {
    return `<${new BigNumber(0.000001).toFormat()}`
  } else if (roundedNumber.isLessThan(0.001)) {
    return precise ? roundUp(value, 6).toFormat() : `<${new BigNumber(0.001).toFormat()}`
  } else {
    return roundUp(value, 3).toFormat()
  }
}

export const divideByWei = (value: BigNumber.Value, decimals?: number) => {
  const bn = new BigNumber(value).div(WEI_PER_TOKEN)
  return decimals ? bn.decimalPlaces(decimals) : bn
}

export const multiplyByWei = (value: BigNumber.Value, decimals?: number) => {
  const bn = new BigNumber(value).times(WEI_PER_TOKEN)
  return decimals ? bn.decimalPlaces(decimals) : bn
}

export function roundDown(
  value: BigNumber.Value,
  decimals: number = 2,
  roundingTolerance: number = 0
): BigNumber {
  if (roundingTolerance) {
    value = new BigNumber(value).decimalPlaces(
      decimals + roundingTolerance,
      BigNumber.ROUND_HALF_DOWN
    )
  }
  return new BigNumber(value).decimalPlaces(decimals, BigNumber.ROUND_DOWN)
}

export function roundUp(
  value: BigNumber.Value,
  decimals: number = 2,
  roundingTolerance: number = 0
): BigNumber {
  if (roundingTolerance) {
    value = new BigNumber(value).decimalPlaces(
      decimals + roundingTolerance,
      BigNumber.ROUND_HALF_DOWN
    )
  }
  return new BigNumber(value).decimalPlaces(decimals, BigNumber.ROUND_UP)
}

export const getCurrencyColor = (currencyType: CURRENCY_ENUM): string => {
  switch (currencyType) {
    case CURRENCY_ENUM.DOLLAR:
      return colors.greenBrand
    case CURRENCY_ENUM.GOLD:
      return colors.goldBrand
  }
}

export const getBalanceColor = (accountBalance: BigNumber): string => {
  if (accountBalance.isGreaterThan(0)) {
    return colors.greenBrand
  }
  if (accountBalance.isLessThan(0)) {
    return colors.warning
  }
  return colors.dark
}
