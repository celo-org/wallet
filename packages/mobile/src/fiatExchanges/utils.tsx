import {
  CurrencyCode,
  DEFAULT_TESTNET,
  IP_ADDRESS_LOCATION_URL_PROD,
  IP_ADDRESS_LOCATION_URL_STAGING,
  PROVIDER_URL_COMPOSER_PROD,
  PROVIDER_URL_COMPOSER_STAGING,
  SIMPLEX_URI,
} from 'src/config'
import { Providers } from 'src/fiatExchanges/ProviderOptionsScreen'
import { providerAvailability } from 'src/flags'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

interface RequestData {
  address: string | null
  digitalAsset: string
  fiatCurrency: string
  fiatAmount: number
}
export interface UserLocation {
  country: string | null
  state: string | null
}

export const fetchProviderWidgetUrl = async (provider: Providers, requestData: RequestData) => {
  const response = await fetch(
    DEFAULT_TESTNET === 'mainnet' ? PROVIDER_URL_COMPOSER_PROD : PROVIDER_URL_COMPOSER_STAGING,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...requestData,
        provider,
        env: DEFAULT_TESTNET === 'mainnet' ? 'production' : 'staging',
      }),
    }
  )

  return response.json()
}

export const fetchUserIpAddress = async () => {
  const response = await fetch(
    DEFAULT_TESTNET === 'mainnet' ? IP_ADDRESS_LOCATION_URL_PROD : IP_ADDRESS_LOCATION_URL_STAGING,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        urlType: 'ip',
        env: DEFAULT_TESTNET === 'mainnet' ? 'production' : 'staging',
      }),
    }
  )

  return response.json()
}

export const isExpectedUrl = (fetchedUrl: string, providerUrl: string) =>
  fetchedUrl.startsWith(providerUrl)

export const openMoonpay = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.MoonPayScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}

export const openRamp = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.RampScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

export const openTransak = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.TransakScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

type ProviderAvailability = typeof providerAvailability
type SpecificProviderAvailability = { [K in keyof ProviderAvailability]: boolean }

type Entries<T> = Array<{ [K in keyof T]: [K, T[K]] }[keyof T]>

export function getProviderAvailability(
  userLocation: UserLocation | undefined
): SpecificProviderAvailability {
  // tslint:disable-next-line: no-object-literal-type-assertion
  const { countryCodeAlpha2, stateCode } = userLocation
    ? { countryCodeAlpha2: userLocation.country, stateCode: userLocation.state }
    : { countryCodeAlpha2: null, stateCode: null }

  const features = {} as SpecificProviderAvailability
  for (const [key, value] of Object.entries(providerAvailability) as Entries<
    ProviderAvailability
  >) {
    if (!countryCodeAlpha2) {
      features[key] = false
    } else {
      if (countryCodeAlpha2 === 'US' && (value as any)[countryCodeAlpha2] !== true) {
        features[key] = stateCode ? (value as any)[countryCodeAlpha2][stateCode] ?? false : false
      } else {
        features[key] = (value as any)[countryCodeAlpha2] ?? false
      }
    }
  }
  return features
}
