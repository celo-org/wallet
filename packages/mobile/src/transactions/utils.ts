import { TokenTransactionType } from 'src/apollo/types'
import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import i18n from 'src/i18n'
import { AddressToDisplayNameType } from 'src/identity/reducer'
import { FeedItem } from 'src/transactions/TransactionFeed'
import { TransferConfirmationCardProps } from 'src/transactions/TransferConfirmationCard'
import { formatFeedSectionTitle, timeDeltaInDays } from 'src/utils/time'

// Groupings:
// Recent -> Last 7 days.
// [Current month] - "July" -> Captures transactions from the current month that aren’t captured in Recent.
// [Previous months] - "June" -> Captures transactions by month.
// [Months over a year ago] — "July 2019" -> Same as above, but with year appended.
// Sections are hidden if they have no items.
export const groupFeedItemsInSections = (feedItems: FeedItem[]) => {
  const sectionsMap: {
    [key: string]: {
      data: FeedItem[]
      daysSinceTransaction: number
    }
  } = {}

  feedItems.reduce((sections, item) => {
    const daysSinceTransaction = timeDeltaInDays(Date.now(), item.timestamp)
    const key =
      daysSinceTransaction <= 7
        ? i18n.t('walletFlow5:feedSectionHeaderRecent')
        : formatFeedSectionTitle(item.timestamp, i18n)
    sections[key] = sections[key] || {
      daysSinceTransaction,
      data: [],
    }
    sections[key].data.push(item)
    return sections
  }, sectionsMap)

  return Object.entries(sectionsMap)
    .sort((a, b) => a[1].daysSinceTransaction - b[1].daysSinceTransaction)
    .map(([key, value]) => ({
      title: key,
      data: value.data,
    }))
}

export const exchangeReviewHeader = (confirmationProps: ExchangeConfirmationCardProps) => {
  const { makerAmount } = confirmationProps
  const isSold = makerAmount.currencyCode === CURRENCIES[CURRENCY_ENUM.GOLD].code
  return isSold ? i18n.t('exchangeFlow9:soldGold') : i18n.t('exchangeFlow9:purchasedGold')
}

export const transferReviewHeader = (
  type: TokenTransactionType,
  confirmationProps: TransferConfirmationCardProps,
  addressToDisplayName: AddressToDisplayNameType
) => {
  let headerText = ''
  const isCeloTx = confirmationProps.amount.currencyCode === CURRENCIES[CURRENCY_ENUM.GOLD].code
  switch (type) {
    case TokenTransactionType.Sent:
      headerText = i18n.t(
        isCeloTx ? 'walletFlow5:transactionHeaderWithdrewCelo' : 'walletFlow5:transactionHeaderSent'
      )
      break
    case TokenTransactionType.EscrowSent:
      headerText = i18n.t('walletFlow5:transactionHeaderEscrowSent')
      break
    case TokenTransactionType.Received:
      if (addressToDisplayName[confirmationProps.address || '']?.isCeloRewardSender) {
        headerText = i18n.t('walletFlow5:transactionHeaderCeloReward')
      }
      headerText = isCeloTx
        ? i18n.t('walletFlow5:transactionHeaderCeloDeposit')
        : i18n.t('walletFlow5:transactionHeaderReceived')
      break
    case TokenTransactionType.EscrowReceived:
      headerText = i18n.t('walletFlow5:transactionHeaderEscrowReceived')
      break
    case TokenTransactionType.VerificationFee:
      headerText = i18n.t('walletFlow5:transactionHeaderVerificationFee')
      break
    case TokenTransactionType.Faucet:
      headerText = i18n.t('walletFlow5:transactionHeaderFaucet')
      break
    case TokenTransactionType.InviteSent:
      headerText = i18n.t('walletFlow5:transactionHeaderInviteSent')
      break
    case TokenTransactionType.InviteReceived:
      headerText = i18n.t('walletFlow5:transactionHeaderInviteReceived')
      break
    case TokenTransactionType.NetworkFee:
      headerText = i18n.t('walletFlow5:transactionHeaderNetworkFee')
      break
  }

  return headerText
}
