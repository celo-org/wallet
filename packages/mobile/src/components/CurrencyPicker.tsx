import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import { Spacing } from '@celo/react-components/styles/styles'
import BigNumber from 'bignumber.js'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { SendEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import { Namespaces } from 'src/i18n'
import { useCurrencyBalance } from 'src/stableToken/hooks'
import { CURRENCIES, Currency, CurrencyInfo, STABLE_CURRENCIES } from 'src/utils/currencies'

export enum CurrencyPickerOrigin {
  Send = 'Send',
  Exchange = 'Exchange',
}

interface Props {
  isVisible: boolean
  origin: CurrencyPickerOrigin
  onCurrencySelected: (currency: Currency) => void
}

function CurrencyOption({
  currency,
  currencyInfo,
  onPress,
}: {
  currency: Currency
  currencyInfo: CurrencyInfo
  onPress: () => void
}) {
  const { t } = useTranslation(Namespaces.sendFlow7)
  const balance = useCurrencyBalance(currency)
  const amount = {
    value: new BigNumber(balance ?? '0'),
    currencyCode: currencyInfo.code,
  }
  return (
    <TouchableOpacity
      style={styles.currencyOptionContainer}
      onPress={onPress}
      testID={`${currencyInfo.code}Touchable`}
    >
      <Text style={styles.optionName}>{t('stableBalance', { token: currencyInfo.code })}</Text>
      <View style={styles.currencyBalanceContainer}>
        <CurrencyDisplay
          style={styles.localBalance}
          amount={amount}
          showLocalAmount={true}
          testID={`Local${currencyInfo.code}Balance`}
        />
        <CurrencyDisplay
          style={styles.currencyBalance}
          amount={amount}
          showLocalAmount={false}
          hideCode={false}
          hideSymbol={true}
          testID={`${currencyInfo.code}Balance`}
        />
      </View>
    </TouchableOpacity>
  )
}

function CurrencyPicker({ isVisible, origin, onCurrencySelected }: Props) {
  const { t } = useTranslation(Namespaces.sendFlow7)
  if (!isVisible) {
    return null
  }
  const onCurrencyPressed = (currency: Currency) => () => {
    ValoraAnalytics.track(SendEvents.token_selected, {
      origin,
      token: CURRENCIES[currency].code,
    })
    onCurrencySelected(currency)
  }
  return (
    <View style={styles.container} testID="CurrencyPickerContainer">
      <View style={styles.background} />
      <View style={styles.contentContainer}>
        <Text style={styles.title}>{t('selectBalance')}</Text>
        {STABLE_CURRENCIES.map((currency, index) => {
          return (
            <>
              {index > 0 && <View style={styles.separator} />}
              <CurrencyOption
                currency={currency}
                currencyInfo={CURRENCIES[currency]}
                onPress={onCurrencyPressed(currency)}
              />
            </>
          )
        })}
      </View>
    </View>
  )
}

CurrencyPicker.navigationOptions = {}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    justifyContent: 'flex-end',
  },
  background: {
    position: 'absolute',
    backgroundColor: colors.modalBackdrop,
    opacity: 0.5,
    width: '100%',
    height: '100%',
  },
  contentContainer: {
    backgroundColor: colors.light,
    opacity: 1,
    width: '100%',
    padding: Spacing.Thick24,
    borderTopRightRadius: Spacing.Regular16,
    borderTopLeftRadius: Spacing.Regular16,
  },
  title: {
    ...fontStyles.h2,
    marginBottom: Spacing.Smallest8,
  },
  currencyOptionContainer: {
    flexDirection: 'row',
    height: 72,
    alignItems: 'center',
  },
  optionName: {
    flex: 1,
    ...fontStyles.regular500,
  },
  currencyBalanceContainer: {
    alignItems: 'flex-end',
  },
  localBalance: {
    ...fontStyles.regular,
  },
  currencyBalance: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  separator: {
    width: '100%',
    height: 1,
    backgroundColor: colors.gray2,
  },
})

export default CurrencyPicker
