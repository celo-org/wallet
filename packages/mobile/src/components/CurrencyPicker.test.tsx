import * as React from 'react'
import { fireEvent, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import CurrencyPicker, { CurrencyPickerOrigin } from 'src/components/CurrencyPicker'
import { Currency } from 'src/utils/currencies'
import { createMockStore } from 'test/utils'

const mockStore = createMockStore({
  stableToken: {
    balance: '10',
    cEurBalance: '20',
  },
})

const onCurrencySelectedMock = jest.fn()

describe('CurrencyPicker', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  function renderPicker(visible: boolean) {
    return render(
      <Provider store={mockStore}>
        <CurrencyPicker
          isVisible={visible}
          origin={CurrencyPickerOrigin.Send}
          onCurrencySelected={onCurrencySelectedMock}
        />
      </Provider>
    )
  }

  it('renders correctly', () => {
    const tree = renderPicker(true)

    expect(tree.queryByTestId('CurrencyPickerContainer')).toBeTruthy()
    expect(tree.queryByTestId('LocalcUSDBalance')).toBeTruthy()
    expect(tree.queryByTestId('cUSDBalance')).toBeTruthy()
    expect(tree.queryByTestId('LocalcEURBalance')).toBeTruthy()
    expect(tree.queryByTestId('cEURBalance')).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('handles the choosing of a currency correctly', () => {
    const { getByTestId } = renderPicker(true)

    fireEvent.press(getByTestId('cUSDTouchable'))
    expect(onCurrencySelectedMock).toHaveBeenLastCalledWith(Currency.Dollar)

    fireEvent.press(getByTestId('cEURTouchable'))
    expect(onCurrencySelectedMock).toHaveBeenLastCalledWith(Currency.Euro)
  })

  it('renders nothing if not visible', () => {
    const { queryByTestId } = renderPicker(false)
    expect(queryByTestId('CurrencyPickerContainer')).toBeFalsy()
  })
})
