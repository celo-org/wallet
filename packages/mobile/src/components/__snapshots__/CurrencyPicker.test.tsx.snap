// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyPicker renders correctly 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "justifyContent": "flex-end",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
  testID="CurrencyPickerContainer"
>
  <View
    style={
      Object {
        "backgroundColor": "#1B392F",
        "height": "100%",
        "opacity": 0.5,
        "position": "absolute",
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderTopLeftRadius": 16,
        "borderTopRightRadius": 16,
        "opacity": 1,
        "padding": 24,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#2E3338",
          "fontFamily": "Jost-Medium",
          "fontSize": 22,
          "lineHeight": 28,
          "marginBottom": 8,
        }
      }
    >
      selectBalance
    </Text>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="cUSDTouchable"
    >
      <View
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 72,
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flex": 1,
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          stableBalance
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                },
              ]
            }
            testID="LocalcUSDBalance"
          >
            
            $
            13.30
          </Text>
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
                Object {
                  "color": "#9CA4A9",
                },
              ]
            }
            testID="cUSDBalance"
          >
            
            10.00
             cUSD
          </Text>
        </View>
      </View>
    </RNGestureHandlerButton>
    <View
      style={
        Object {
          "backgroundColor": "#EDEEEF",
          "height": 1,
          "width": "100%",
        }
      }
    />
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="cEURTouchable"
    >
      <View
        accessible={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 72,
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flex": 1,
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          stableBalance
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                },
              ]
            }
            testID="LocalcEURBalance"
          >
            
            $
            40.00
          </Text>
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
                Object {
                  "color": "#9CA4A9",
                },
              ]
            }
            testID="cEURBalance"
          >
            
            20.00
             cEUR
          </Text>
        </View>
      </View>
    </RNGestureHandlerButton>
  </View>
</View>
`;
